### DISCUSSION TITLE
error: undefined reference to 'vtable for S2Cap' when build doris 2.0.1.1 source code

### DISCUSSION THREAD

**Discussion by fengyeGit** [apache/doris#25139] [A - General / Q&A]
has anyone encounter this problem when build doris source code, detailed infomation as follows:
 Linker GNU gold is available, change linker to GNU gold
-- Some symbols from glibc will be replaced for compatibility
-- Compiler: GNU-11.3.0
-- CXX Standard: 20
-- C Standard: 17
-- CXX Flags:   -O3 -DNDEBUG
-- C Flags:   -O3 -DNDEBUG
-- CC Flags:
[3/3] Linking CXX executable src/service/doris_be
FAILED: src/service/doris_be
: && /usr/local/bin/g++ -O3 -DNDEBUG  -O3 -DNDEBUG -fuse-ld=gold   -Wl,--export-dynamic -rdynamic src/glibc-compatibility/CMakeFiles/glibc-compatibility-explicit.dir/musl/getrandom.c.o src/glibc-compatibility/CMakeFiles/glibc-compatibility-explicit.dir/memcpy/memcpy_x86_64.cpp.o src/service/CMakeFiles/doris_be.dir/doris_main.cpp.o -o src/service/doris_be  -Wl,--start-group  src/agent/libAgent.a  src/common/libCommon.a  src/exec/libExec.a  src/exprs/libExprs.a  src/gutil/libGutil.a  src/io/libIO.a  src/olap/libOlap.a  src/olap/rowset/libRowset.a  src/runtime/libRuntime.a  src/service/libService.a  src/udf/libUdf.a  src/util/libUtil.a  src/gen_cpp/libDorisGen.a  src/http/libWebserver.a  src/geo/libGeo.a  src/vec/libVec.a  src/pipeline/libPipeline.a  -Wl,--end-group  -Wl,--start-group  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libbacktrace.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/librocksdb.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libsasl2.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libs2.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libsnappy.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libboost_date_time.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libthrift.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libthriftnb.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libglog.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libre2.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libhs.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/gperftools/lib/libprofiler.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/liblz4.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libevent.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libevent_pthreads.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libidn.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libgsasl.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libcurl.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libz.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libbz2.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libgflags.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libbrpc.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libprotobuf.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libssl.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libcrypto.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libleveldb.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libbitshuffle.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libroaring.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libfmt.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libjemalloc_doris.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libjemalloc.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libbrotlicommon.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libbrotlidec.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libbrotlienc.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libzstd.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libarrow.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libparquet.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libodbc.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libcctz.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libminizip.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libopentelemetry_common.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libopentelemetry_exporter_zipkin_trace.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libopentelemetry_resources.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libopentelemetry_version.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libopentelemetry_exporter_ostream_span.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libopentelemetry_trace.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libopentelemetry_http_client_curl.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libopentelemetry_exporter_otlp_http.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libopentelemetry_exporter_otlp_http_client.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libopentelemetry_otlp_recordable.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libopentelemetry_proto.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libaws-cpp-sdk-s3.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libaws-cpp-sdk-core.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libaws-cpp-sdk-transfer.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libaws-checksums.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libaws-c-io.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libaws-c-event-stream.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libaws-c-common.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libaws-c-cal.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libs2n.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libaws-c-s3.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libaws-c-auth.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libaws-crt-cpp.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libaws-c-compression.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libaws-c-http.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libaws-c-mqtt.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libaws-cpp-sdk-s3-crt.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/librdkafka++.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/librdkafka.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libxml2.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/liblzma.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libsimdjson.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/hadoop_hdfs/native/libhdfs.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libkrb5support.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libkrb5.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libcom_err.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libgssapi_krb5.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libk5crypto.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libunwind.a  bin/liborc.a  bin/libic.a  bin/libclucene-core-static.a  bin/libclucene-shared-static.a  bin/libclucene-contribs-lib.a  -Wl,--end-group  -static-libstdc++  -static-libgcc  -lstdc++fs  -lresolv  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libjemalloc_doris.a  bin/libglibc-compatibility.a  -lrt  -l:libbfd.a  -liberty  -lc  -lm  -ldl  -pthread  src/geo/libgeo_type.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_cord.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_cordz_info.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_cord_internal.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_cordz_functions.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_cordz_handle.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_hash.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_city.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_bad_variant_access.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_low_level_hash.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_raw_hash_set.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_bad_optional_access.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_hashtablez_sampler.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_exponential_biased.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_synchronization.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_stacktrace.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_graphcycles_internal.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_symbolize.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_malloc_internal.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_debugging_internal.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_demangle_internal.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_time.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_civil_time.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_time_zone.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_str_format_internal.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_strings.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_strings_internal.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_base.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_spinlock_wait.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_int128.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_throw_delegate.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_raw_logging_internal.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib64/libabsl_log_severity.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libprotobuf.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libz.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libsnappy.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/liblz4.a  /home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/lib/libzstd.a  bin/libclucene-core-static.a  bin/libclucene-shared-static.a  -lpthread  -lrt && :

/home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/include/s2/s2cap.h:62: error: undefined reference to 'vtable for S2Cap'
/usr/bin/ld.gold: the vtable symbol may be undefined because the class is missing its key function
/home/yecg/apache-doris-2.0.1.1-src/be/src/geo/geo_types.cpp:544: error: undefined reference to 'S2Cap::Decode(Decoder*)'
/home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/include/s2/s2cap.h:278: error: undefined reference to 'S2::IsUnitLength(Vector3<double> const&)'
/home/yecg/apache-doris-2.0.1.1-src/be/src/geo/geo_types.cpp:59: error: undefined reference to 'S2LatLng::S2LatLng(Vector3<double> const&)'
/home/yecg/apache-doris-2.0.1.1-src/be/src/geo/geo_types.cpp:196: error: undefined reference to 'doris::WktParse::parse_wkt(char const*, unsigned long, doris::GeoShape**)'
/home/yecg/apache-doris-2.0.1.1-src/be/src/geo/geo_types.cpp:74: error: undefined reference to 'S2LatLng::ToPoint() const'
/home/yecg/apache-doris-2.0.1.1-src/be/src/geo/geo_types.cpp:74: error: undefined reference to 'S2LatLng::ToPoint() const'
/home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/include/s2/s2earth.h:249: error: undefined reference to 'S2LatLng::GetDistance(S2LatLng const&) const'
/home/yecg/apache-doris-2.0.1.1-src/be/src/geo/geo_types.cpp:360: error: undefined reference to 'S2LatLng::GetDistance(S2LatLng const&) const'
/home/yecg/apache-doris-2.0.1.1-src/be/src/geo/geo_types.cpp:360: error: undefined reference to 'S2LatLng::GetDistance(S2LatLng const&) const'
/home/yecg/apache-doris-2.0.1.1-src/be/src/geo/geo_types.cpp:360: error: undefined reference to 'S2LatLng::GetDistance(S2LatLng const&) const'
/home/yecg/apache-doris-2.0.1.1-src/be/src/geo/geo_types.cpp:385: error: undefined reference to 'S2Earth::GetInitialBearing(S2LatLng const&, S2LatLng const&)'
/home/yecg/apache-doris-2.0.1.1-src/be/src/geo/geo_types.cpp:387: error: undefined reference to 'S2Earth::GetInitialBearing(S2LatLng const&, S2LatLng const&)'
/home/yecg/apache-doris-2.0.1.1-src/be/src/geo/geo_types.cpp:74: error: undefined reference to 'S2LatLng::ToPoint() const'
/home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/include/s2/s2cap.h:241: error: undefined reference to 'vtable for S2Cap'
/usr/bin/ld.gold: the vtable symbol may be undefined because the class is missing its key function
/home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/include/s2/s2cap.h:241: error: undefined reference to 'S1ChordAngle::S1ChordAngle(S1Angle)'
/home/yecg/apache-doris-2.0.1.1-src/thirdparty/installed/include/s2/s2cap.h:278: error: undefined reference to 'S2::IsUnitLength(Vector3<double> const&)'
/home/yecg/apache-doris-2.0.1.1-src/be/src/geo/geo_types.cpp:556: error: undefined reference to 'S2Cap::GetArea() const'
/home/yecg/apache-doris-2.0.1.1-src/be/src/geo/geo_types.cpp:499: error: undefined reference to 'S2Polygon::GetArea() const'
/home/yecg/apache-doris-2.0.1.1-src/be/src/geo/geo_types.cpp:597: error: undefined reference to 'doris::toBinary::geo_tobinary(doris::GeoShape*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)'
/home/yecg/apache-doris-2.0.1.1-src/be/src/geo/geo_types.cpp:74: error: undefined reference to 'S2LatLng::ToPoint() const'
/home/yecg/apache-doris-2.0.1.1-src/be/src/geo/geo_types.cpp:408: error: undefined reference to 'S2Polyline::Encode(Encoder*) const'
/home/yecg/apache-doris-2.0.1.1-src/be/src/geo/geo_types.cpp:410: error: undefined reference to 'Encoder::~Encoder()'
/home/yecg/apache-doris-2.0.1.1-src/be/src/geo/geo_types.cpp:432: error: undefined reference to 'S2Polygon::Encode(Encoder*) const'
/home/yecg/apache-doris-2.0.1.1-src/be/src/geo/geo_types.cpp:434: error: undefined reference to 'Encoder::~Encoder()'
/home/yecg/apache-doris-2.0.1.1-src/be/src/geo/geo_types.cpp:537: error: undefined reference to 'S2Cap::Encode(Encoder*) const'

**Comment by LemonLiTree** [comment]
How to compile the source code？LDB？
